gcstorage/gcstorage.go

```diff
@@ -0,0 +1,253 @@
+package gcstorage
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"time"
+
+	"cloud.google.com/go/storage"
+	"github.com/goware/cachestore"
+	"google.golang.org/api/iterator"
+	"google.golang.org/api/option"
+)
+
+var ErrNotSupported = fmt.Errorf("not supported")
+
+type cacheObject[T any] struct {
+	Object    T         `json:"object"`
+	ExpiresAt time.Time `json:"expires_at"`
+}
+
+type GCStorage[V any] struct {
+	keyPrefix string
+
+	defaultKeyExpiry time.Duration
+
+	client       *storage.Client
+	bucketHandle *storage.BucketHandle
+}
+
+func Backend(cfg *Config, opts ...cachestore.StoreOptions) cachestore.Backend {
+	options := cachestore.ApplyOptions(opts...)
+	cfg.StoreOptions = options
+	return cfg
+}
+
+func NewWithBackend[V any](backend cachestore.Backend, opts ...cachestore.StoreOptions) (cachestore.Store[V], error) {
+	cfg, ok := backend.(*Config)
+	if !ok {
+		return nil, fmt.Errorf("cachestore/redis: invalid backend config supplied")
+	}
+	for _, opt := range opts {
+		opt.Apply(&cfg.StoreOptions)
+	}
+	return New[V](cfg, cfg.StoreOptions)
+}
+
+func New[V any](cfg *Config, opts ...cachestore.StoreOptions) (cachestore.Store[V], error) {
+	for _, opt := range opts {
+		opt.Apply(&cfg.StoreOptions)
+	}
+
+	var gcOpts []option.ClientOption
+	if cfg.Credentials != nil {
+		gcOpts = append(gcOpts, option.WithCredentials(cfg.Credentials))
+	}
+
+	client, err := storage.NewClient(context.Background(), gcOpts...)
+	if err != nil {
+		return nil, err
+	}
+
+	return &GCStorage[V]{
+		keyPrefix:        cfg.KeyPrefix,
+		defaultKeyExpiry: cfg.DefaultKeyExpiry,
```

Comment:
IMHO we should set some default expiry on zero? Something like one day? same as on redis client: https://github.com/goware/cachestore/blob/master/redis/redis.go#L17

It sounds weird with gcloud storage client. But this mainly is "caching" library and not library for saving objects to gcloud storage.

Reply:
I have set the defaut expiry to 1 day as suggested. In some cases, we may want to cache stuff forever, but in this case, we can set time max.

We were discussing using this for node-gateway to cache data that do not change. So we do not query origin nodes too much. In my particular case, I use it in ``CoinGeckoPriceSource`` to cache prices, so when I'm re-running the data pipeline I do not need to hammer CoinGecko all the time.